### ISSUE TITLE
Automatically switch resolution depending on cache condition

### ISSUE THREAD

**Issue by anishmittal2020** [mpv-player/mpv#9074]
### Expected behavior of the wanted feature

Sometimes the internet is not reliable, it would be nice if it is possible to automatically change resolution when cache speed is slower than playing speed (say, continuously for 5 seconds) and when the difference between max cache position and current playing position is less than 30 seconds.

And may be it could step up resolution if the internet has good connection.

**Comment by LaserEyess** [comment]
I don't think this is possible, certainly not in the way it works on sites like youtube or netflix. They use a ton of client side javascript utilizing private APIs to send requests and monitor network bandwidth to determine the optimal bitrate/resolution. None of this is available to non-browser clients, certainly not to mpv or ffmpeg. Even if this was possible, it's something for both youtube-dl and ffmpeg to support. youtube-dl would need to be able to return multiple URLs for the different resolutions and ffmpeg would need a way to emulate a browser client to make the right POSTs get the sites to return the correct JSON chunks. Until youtube-dl and ffmpeg support this, mpv can't do anything.

This *may* be possible with a user script, but even then it would not be what you want. It would be a script that continuously polls your cache and uses some algorithm to guess network speed, because that's not available to mpv, and then retry the connection with a lower or higher resolution. This would certainly not be "seamless" like it is for youtube, and would be a manual restart of the stream with new parameters.


TL;DR: You're asking for something that's nearly impossible for mpv to do in a way that would work well. These things are usually, by design, not meant for 3rd party tools.

**Comment by ValZapod** [comment]
> requests and monitor network bandwidth

No, the main parameter is Google Cache availability in particular AS and its propogation and global troughput. Also, it is latency that plays a big role, little latency higher resolution. If everybody was forcing HDR 8K 60 fps it would have crushed google AS.

**Comment by gerion0** [comment]
YoutubeDL provides a bunch of available resolutions already. Check `youtube-dl -F <youtube-link>`. Isn't it possible that mpv only checks if the cache is filling fast enough (this means faster or equal to the playing speed) and switches to a lower resolution, if this is not the case?

**Comment by Dudemanguy** [comment]
As explained in the comments, this feature isn't really practical for mpv.
